### PR TITLE
Request for read interest when channel is unwritable while HttpClient.send(Mono) is used

### LINKED ISSUE #2825: Misleading reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
When we use spring boot's web client with netty connection provider or http client and call http POST with very large JSON payload, it throws error reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response.
But actually this is misleading instead of propagating real http 400 BAD_REQUEST. I went into rabbit hole of this guide: https://projectreactor.io/docs/netty/release/reference/index.html#faq.connection-closed and still could not figure out until I kept increasing load on the REST api.

I was able to reproduce this locally running spring boot app without any load balancer or proxy in between.

## Expected Behavior

client should propagate actual http 400 BAD REQUEST error instead of masking with PrematureCloseException.
## Actual Behavior

throws error reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
## Steps to Reproduce

Run the unit test:
https://github.com/himanshunp/spring-web-client/blob/master/src/test/java/com/parmar/himanshu/spring/webclient/ControllerTest.java#L37
If you look at the three parameters, the last one tests with very large json payload and reproduces above error. There should be a better way to capture http 400 and response body.

## Possible Solution

I was able to track down this by using X-Request-ID header. The documentation should add this as a debugging step.
## Your Environment

* Reactor version(s) used: 3.5.6
* Other relevant libraries versions (eg. `netty`, ...): `reactor-netty-core`: 1.1.7
* JVM version (`java -version`): 20
* OS and version (eg. `uname -a`): macOS Monterey 12.6.6 Darwin himanshus-mbp.lan 21.6.0 Darwin Kernel Version 21.6.0: Mon Apr 24 21:10:53 PDT 2023; root:xnu-8020.240.18.701.5~1/RELEASE_X86_64 x86_64

### PR DESCRIPTION
**Pull request by pderop**
**Motivation:**

When using `HttpClient.send` with a `Mono` publisher and `plain POST HTTP/1.1`, a problem may occur if the channel becomes unwritable during request flushing. If the remote server sends an early response (e.g., 400 bad request) and closes the connection without reading the request post data, the client might fail with a "Connection prematurely closed BEFORE response" error instead of reporting the actual early server response (e.g., 400 bad request). 

This issue arises because, in the case of a Mono<ByteBuf> publisher and HTTP/1.1 plain, the channel read interest is not enabled when the channel becomes unwritable, preventing the client from reading the response and leading to "Connection prematurely closed BEFORE response" errors.

For example, in GH #2825, the user has provided a reproducer example, where the server (port 8000) sends a 400 bad request to the client which is still writing a large POST HTTP/1.1 request data (the client is often blocked while writing because of TCP flow control). In the following wireshark, the frame 602 is the 400 bad request returned to the user. Notice that the 400 bad request has been sent in two frames (in frame 598, the client first gets the 400 bad request headers in chunk, followed by the body, and in frame 602, we get the last zero-length chunk header that is delimiting the end of the message):

Screenshot 2023-07-26 at 21 15 56

Then in frame 603, the server receives the TCP/ACK highlighted with a `[TCP Window Full]` info, meaning that the client is using the full capacity of the TCP flow, limited by the server's receive window which is full.

Finally, a TCP/RST is then sent to the client in frame 767, but the client has not requested for read() interest, so it misses the 400 bad request, and then aborts with `Connection prematurely closed BEFORE response` errors:

Screenshot 2023-07-26 at 21 16 21

By requesting for read interests while writing the big POST data request, we can alleviate the issue and be able to report the 400 bad request to the user instead of misleading `Connection prematurely closed BEFORE response`

**Caution:** This PR may alleviate the issue, but sometimes, it may not avoid the problem because TCP/RST is not graceful and some packets may not be handled by the client. For example, if the 400 bad request is sent in chunked encoding and is segmented in 3 frames (1st with the headers, 2nd with the the body, and the 3rd with the last zero-length chunk), then the last 3rd frame with the last chunk EOF may be missed, even if we see it from wireshark on the client machine ...). In this case, we may get a **Connection prematurely closed DURING response** error on the client. This we sometimes won't be able to avoid it (because of the nature of the TCP/RST).  

For example, with curl, we usually always get the 400 bad request, but if we manage to let the server send the 400 bad request in three frames (the headers, the body, and the last chunk), then sometimes curls will fail like this:
```
curl -v --header "Expect:"  -H "Content-Type: text/plain" -d @sample_request.json http://xxxx.xxxx.xxxx.xxxx:8000/payload-size
> POST /large-payload HTTP/1.1
> Host: xxxx.xxxx.xxxx.xxxx:8000
> User-Agent: curl/7.88.1
> Accept: */*
> Content-Type: text/plain
> Content-Length: 1561039
> 
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

But most of the time, curl is able to get the 400 bad request response:

In the context of Tomcat (used by the reproducer, the 400 bad request is sent in two frames: first for the headers+body, second one for the last chunk header.

**Modifications:**

Requesting for read interest when `HttpClient.send` method is used with a `Mono` publisher seems to fix the issue, and we can now get the early 400 bad request server response instead of the unexpected `Connection prematurely closed BEFORE response` error. 

Now, read interest could be systematically enabled, like this in the HttpOperations.send(Publisher) method:

```
	public NettyOutbound send(Publisher<? extends ByteBuf> source) {
		if (!channel().isActive()) {
			return then(Mono.error(AbortedException.beforeSend()));
		}
		if (source instanceof Mono) {
			return new PostHeadersNettyOutbound(((Mono<ByteBuf>) source)
					.flatMap(b -> {
						if (markSentHeaderAndBody(b)) {
							HttpMessage msg = prepareHttpMessage(b);

							try {
								afterMarkSentHeaders();
							}
							catch (RuntimeException e) {
								ReferenceCountUtil.release(b);
								return Mono.error(e);
							}
->                                                      channel.read();
							return FutureMono.from(channel().writeAndFlush(msg));
						}
						return FutureMono.from(channel().writeAndFlush(b));
					})
					.doOnDiscard(ByteBuf.class, ByteBuf::release), this, null);
		}
		return super.send(source);
	}

```

Instead of doing a read() systematically, this PR uses a different approach and only requests for channel `read()` if the channel becomes unwritable: The `ChannelOperationsHandler` is now overriding the `channelWritabilityChanged `callback, which delegates to any registered `ChannelOperation` (there is also an empty default `onWritabilityChanged` method that has been added in ChannelOperations). So, the `onWritabilityChanged` is only implemented by the HttpClientOperations class, and when the channel becomes unwritable, it will then invoke `channel().read()`,  but  only if the request has been sent using `HttpClient.send(Mono<ByteBuf>)`, and `plain HTT/1.1` is used. I do not think it is needed to do the same when `HTTPS`, `H2C`, `HTTP2`, or `WebSocket` are used.

In order to detect if `HttpClient.send(Mono<ByteBuf>)` has been used, the PR relies on the `HttpOperations.hasSentBody()` new method. Please check the important javadoc done on top of `HttpClientOperations.onWritabilityChanged()`

There is a last issue: in `AbstractHttpClientMetricsHandler`, since now it's possible that we receive an early response before the corresponding request has been fully flushed, care must be taken, because once `channelRead()` receives a full response, it will call `recordRead()`, and after it will call **reset(),**. So that's a problem if the promise listener of the `write` method completes later, because it will then call `recordWrite()` but at this point, all class fields will have already been  cleared by the `reset()` previously called by `channelRead()`.
To handle such problem: some sequence numbers are now used in order to let the channelRead() detect if the full write has not yet completed when a full response is received. In this case channelRead() will call itself recordWrite() on behalf of the `write` method, and then `recordRead()` is called, and then `reset().`

Added two tests:

- Added HttpClientWithTomcatTest.testIssue2825 which is reproducing the similar scenario from #2825, where tomcat is returning an early 400 bad request response and closes the connection while the client is writing a large POST http request. The test checks both usage of `HttpClient.send(Flux<ByteBuf>)` and `HttpClient.send(Mono<ByteBuf>)`
- Added HttpClientTests.testIssue2825 which validates we have no problems when using HTTP 1.1 secure, H2C, or HTTP2

Fixes #2825

### CODE REVIEW

**Comment by pderop** [comment]
I don't think the windows test failures is related to this PR.

**Comment by pderop** [comment]
@violetagg , 

When you will be available (this is not urgent), can you take a look ?
thanks.

**Comment by pderop** [comment]
I have updated the HttpClientTest in order to also test HTTP/1.1 plain protocol.

**Comment by pderop** [comment]
@violetagg , 

the checks are failing on windows , let me verify that before reviewing ...

**Comment by pderop** [comment]
rebased on top of 1.0.x, in order to pick up #2892

**Comment by pderop** [comment]
After some research, it appears that the tests from this PR that were unstable so far were using a reactor netty http server. But this is a different use case than the reproducer project provided from  #2825, which was based on Tomcat.

When a request is aborted, Tomcat continues to read request body bytes up to two mega bytes once a final response (400) is sent, and this allows the reactor netty client to have more times to be able to see the 400 bad request (using the patch from this PR). See `maxSwallowSize` in [Tomcat documentation](https://tomcat.apache.org/tomcat-8.0-doc/config/http.html).

Now, when using reactor netty http server, the issue is that the connection is closed right after the 400 bad request is sent, and on localhost, the tests may be unstable because TCP/RST is sent to the client, which then may miss the 400 bad request.

I have removed all reactor netty based tests, and I have only left the Tomcat test that is using HTTP/1.1 plain, and the tests are now stable. 

Turning this PR to ready for review (in the latest checks, I don't think that the errors from the Windows matrix is related).

**Comment by violetagg** [comment]
Please fix the `checkstyle` warning

```
> Task :reactor-netty-http:checkstyleMain
/home/runner/work/reactor-netty/reactor-netty/reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientWithTomcatTest.java:75: warning: [UnnecessaryParentheses] These grouping parentheses are unnecessary; it is unlikely the code will be misinterpreted without them
	private static final byte[] PAYLOAD = String.join("", Collections.nCopies((TomcatServer.PAYLOAD_MAX) + (1024 * 1024), "X"))
	                                                                          ^
    (see https://errorprone.info/bugpattern/UnnecessaryParentheses)
  Did you mean 'private static final byte[] PAYLOAD = String.join("", Collections.nCopies(TomcatServer.PAYLOAD_MAX + (1024 * 1024), "X"))'?
```

File: reactor-netty-core/src/main/java/reactor/netty/channel/ChannelOperationsHandler.java [hunk 1]
```diff
@@ -149,6 +149,15 @@ final public void exceptionCaught(ChannelHandlerContext ctx, Throwable err) {
 		}
 	}
 
+	@Override
+	final public void channelWritabilityChanged(ChannelHandlerContext ctx) {
+		ChannelOperations<?, ?> ops = ChannelOperations.get(ctx.channel());
+		if (ops != null) {
+			ops.onWritabilityChanged();
+		}
+		ctx.fireChannelWritabilityChanged();
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-core/src/main/java/reactor/netty/channel/ChannelOperationsHandler.java]: thanks, fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/457a394c761b0caf34d42efde9f07fae645922a2

File: reactor-netty-http/src/main/java/reactor/netty/http/HttpOperations.java [hunk 2]
```diff
@@ -335,6 +335,16 @@ protected final boolean markSentBody() {
 		return HTTP_STATE.compareAndSet(this, HEADERS_SENT, BODY_SENT);
 	}
 
+	/**
+	 * Has Body been sent
+	 *
+	 * @return true if body have been sent
+	 * @since 1.0.35
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-http/src/main/java/reactor/netty/http/HttpOperations.java]: fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/14eee2158383555bd16ce6b94ec377e81aff14e6 and in https://github.com/reactor/reactor-netty/pull/2864/commits/c7ac97fdd6a801fcd35582c4aee25939d4c207f4

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientOperations.java [hunk 3]
```diff
@@ -524,6 +525,46 @@ else if (version.equals(HttpVersion.HTTP_1_1)) {
 		throw new IllegalStateException(version.protocolName() + " not supported");
 	}
 
+	/**
+	 * React on channel unwritability event while the http client request is being written.
+	 *
+	 * <p>When using plain HTTP/1.1 and {@code HttpClient.send(Mono<ByteBuf>)}, if the socket becomes unwritable while writing,
+	 * we need to request for reads. This is necessary to read any early server responses, such as a 400 bad request followed
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientOperations.java]: fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/21af63e51074b0f5c7f22b4921b68b51526e9cc9

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientOperations.java [hunk 4]
```diff
@@ -524,6 +525,46 @@ else if (version.equals(HttpVersion.HTTP_1_1)) {
...
+	 *
+	 * <p>See GH-2825 for more info
+	 */
+	@Override
+	protected void onWritabilityChanged() {
+		if (!channel().isWritable() && !channel().config().isAutoRead() &&
+				hasSentBody() &&
+				!isSecure &&
+				!(channel() instanceof Http2StreamChannel) &&
+				!isWebsocket()) {
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientOperations.java]: fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/082de3849feb1efb370bd5b568d92675f61aae79

File: reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientWithTomcatTest.java [hunk 5]
```diff
@@ -317,11 +334,139 @@ void contentHeader() {
...
+
+		return Stream.of(
+				Arguments.of(Named.of("postMono", postMono), Named.of("bytes", PAYLOAD.length)),
+				Arguments.of(Named.of("postFlux", postFlux), Named.of("bytes", PAYLOAD.length))
+		);
+	}
+
+	@ParameterizedTest
+	@MethodSource("testIssue2825Args")
+	void testIssue2825(@Nullable Supplier<Publisher<ByteBuf>> payload, long bytesToSend) {
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientWithTomcatTest.java]: fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/711105f651f2bcbae234ecaf5f4f3dea9354b9bf

File: reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientWithTomcatTest.java [hunk 6]
```diff
@@ -317,11 +334,139 @@ void contentHeader() {
...
+		AtomicReference<SocketAddress> serverAddress = new AtomicReference<>();
+		HttpClient client = HttpClient.create()
+				.port(getPort())
+				.wiretap(false)
+				.metrics(true, ClientMetricsRecorder::reset)
+				.doOnConnected(conn -> {
+					serverAddress.set(conn.address());
+				});
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientWithTomcatTest.java]: fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/4787ffc546c02e87b33b78de3d24a564aac15292

File: reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientWithTomcatTest.java [hunk 7]
```diff
@@ -317,11 +334,139 @@ void contentHeader() {
...
+
+		StepVerifier.create(client
+				.headers(hdr -> hdr.set("Content-Type", "text/plain"))
+				.post()
+				.uri("/payload-size")
+				.send(payload.get())
+						.response((r, buf) -> buf.aggregate().asString()
+								.zipWith(Mono.just(r))))
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientWithTomcatTest.java]: fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/3f1370d2af9655264d2e5417e3668bb025a8a186

**Comment by pderop** [comment]
Fixed checkstyle warning in https://github.com/reactor/reactor-netty/pull/2864/commits/69c0b1489a399c87c9da6a309c953779f117f710

File: reactor-netty-core/src/main/java/reactor/netty/channel/ChannelOperations.java [hunk 8]
```diff
@@ -544,6 +544,14 @@ protected String asDebugLogMessage(Object o) {
 		return o.toString();
 	}
 
+	/**
+	 * React on  Channel writability change.
+	 *
+	 * @since 1.0.35
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-core/src/main/java/reactor/netty/channel/ChannelOperations.java]: fixed in https://github.com/reactor/reactor-netty/pull/2864/commits/c7ac97fdd6a801fcd35582c4aee25939d4c207f4

**Comment by pderop** [comment]
I have applied your feedbacks (thanks), can you check ?

**Review by violetagg — APPROVED** [review]

**Comment by pderop** [comment]
@violetagg , thanks for the review.

**Pull request merged**